File: feature_finders/feature_finder.py
#!/usr/bin/env python3

# GFF Feature finder
#
# Author: Tyler Aprati
#
# This program finds genes between a coordinate range within a gff file
# Example usage:
# featurefinder.py -i input.gff -o chromosome -s start_index -e end_index
#

import sys, argparse
from argparse import RawTextHelpFormatter

print()

parser = argparse.ArgumentParser(description=
                                 'This program is written for execution in the Python (version 3) language.\n'+
                                 'This program filters a gff file by criteria.\n',
                                 formatter_class=RawTextHelpFormatter)

requiredParam = parser.add_argument_group('required parameters')
requiredParam.add_argument('-i', type=str, metavar='infile', required=True, help='Name of input file')
requiredParam.add_argument('-o', type=str, metavar='outfile', required=True, help='Name for output file')
requiredParam.add_argument('-c', type=str, metavar='chromosome', required=True, help='Chromosome of interest')
requiredParam.add_argument('-t', type=str, metavar='feature_type', required=True, help='Feature Type')
requiredParam.add_argument('-s', type=int, metavar='start_index', required=True, help='Start index for search range')
requiredParam.add_argument('-e', type=int, metavar='end_index', required=True, help='End index for search range')

args = parser.parse_args()

# Load in other inputs
filename = args.i
output = args.o
chrom = args.c
type = args.t
s_index = args.s
e_index = args.e

# Open input and output files
file = open(filename,'r')
output = open(output,'w')

# Print update
print("Searching through",filename)

# Find features
count = 0
while True:
    l = file.readline()
    if l == '':
        break
    line = l.split("\t")
    if line[0] == chrom:
        if line[2] == type:
            if (int(line[3]) > s_index) and (int(line[3]) < e_index):
                output.write(l)
                count += 1

print(count,"Features Found")

output.close()

File: feature_finders/feature_finder.R
#!/usr/bin/env Rscript
# 
# Feature Finder
#

# Read in args
args = commandArgs(trailingOnly = TRUE)

# Assign args
input = args[1]
chrom = args[2]
type = args[3]
start = as.integer(args[4])
end = as.integer(args[5])

# Check for args
if (length(args) != 5) {
  cat("Usage: feature_finder.R input chromosome feature_type start_index end_index\n")
  stop("Invalid input. Not enough arguments!", call.=FALSE)
}

# Load the library.
suppressPackageStartupMessages(library(dplyr))

# Load in data
data <- read.delim(input, header=F, comment.char="#")

# Filter data by parameters
genes_of_interest <- data %>%
  filter(V1 == chrom, 
         V3 == type,
         (V4 >= start & V4 <= end) | (V5 >= start & V5 <= end) | (V4 <= start & V5 >= end)
         )

# Write genes of interest into output file
write.table(genes_of_interest, 
            file = "interesting_genes_R.out", 
            sep = "\t", 
            row.names = FALSE,
            col.names = FALSE, 
            quote = FALSE)

